swap/simulations_test.go

```diff
@@ -0,0 +1,864 @@
+		reflect.TypeOf(testMsgBigPrice{}): {
+			Value:   DefaultPaymentThreshold + 1,
+			PerByte: false,
+			Payer:   protocols.Sender,
+		},
+	}
+}
+
+// Price returns the price for a (test) message
+func (tp *testPrices) Price(msg interface{}) *protocols.Price {
+	t := reflect.TypeOf(msg).Elem()
+	return tp.priceMatrix[t]
+}
+
+// testService encapsulates objects needed for the simulation
+type testService struct {
+	swap  *Swap
+	spec  *protocols.Spec
+	peers map[enode.ID]*testPeer
+}
+
+func newTestService() *testService {
+	return &testService{
+		peers: make(map[enode.ID]*testPeer),
+	}
+}
+
+// testPeer is our object for the test protocol; we can use it to handle our own messages
+type testPeer struct {
+	*protocols.Peer
+}
+
+// handle our own messages; we don't need to do anything (yet), we only
+// want messages to be sent and received, and we need this function for the protocol spec
+func (tp *testPeer) handleMsg(ctx context.Context, msg interface{}) error {
+	return nil
+}
+
+// newSimServiceMap creates the `ServiceFunc` map for node initialization.
+// The trick we need to apply is that we need to create a `SimulatedBackend`
+// with all accounts for every simulation node pre-loaded with "funds".
+// To do this, we pass a `swapSimulationParams` object to this function,
+// which contains the shared objects needed to initialize the `SimulatedBackend`
+func newSimServiceMap(params *swapSimulationParams) map[string]simulation.ServiceFunc {
+	simServiceMap := map[string]simulation.ServiceFunc{
+		// we need the bzz service in order to build up a kademlia
+		"bzz": func(ctx *adapters.ServiceContext, bucket *sync.Map) (node.Service, func(), error) {
+			addr := network.NewAddr(ctx.Config.Node())
+			hp := network.NewHiveParams()
+			hp.Discovery = false
+			config := &network.BzzConfig{
+				OverlayAddr:  addr.Over(),
+				UnderlayAddr: addr.Under(),
+				HiveParams:   hp,
+			}
+			kad := network.NewKademlia(addr.Over(), network.NewKadParams())
+			return network.NewBzz(config, kad, nil, nil, nil, nil, nil), nil, nil
+		},
+		// and we also use a swap service
+		"swap": func(ctx *adapters.ServiceContext, bucket *sync.Map) (s node.Service, cleanup func(), err error) {
+			// every simulation node has an instance of a `testService`
+			ts := newTestService()
+			// balance is the interface for `NewAccounting`; it is a Swap
+			balance := params.swaps[params.count]
+			// every node is a different instance of a Swap and gets a different entry in the map
+			params.count++
+			// to create the accounting, we also need a `Price` instance
+			prices := &testPrices{}
+			// create the matrix of test prices
+			prices.newTestPriceMatrix()
+			ts.spec = newTestSpec()
+			// create the accounting instance and assign to the spec
+			ts.spec.Hook = protocols.NewAccounting(balance, prices)
+			ts.swap = balance
+			// deploy the accounting to the `SimulatedBackend`
+			testDeploy(context.Background(), balance.backend, balance)
+			params.backend.Commit()
```

Comment:
commit already done inside testDeploy

Reply:
Indeed, but it was the "wrong" one, there it used a global `testBackend`, which is not the one we used here, causing actual havoc.

So I also tackled https://github.com/ethersphere/swarm/issues/1741 in this PR, which fixes the problem, and now `testDeploy` is correctly committed there.